From 26af68d8647749abac7c6430c1d08ec1bfae4de2 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Mon, 15 Jun 2026 02:43:40 +0000
Subject: [PATCH] Fix Use-After-Free in NXP NFC HAL timer teardown

A vulnerability existed where a compromised NFC Controller could manipulate POSIX timer delays during the INfc::close() teardown. This allowed a detached SIGEV_THREAD timer callback to outlive its deletion and access a freed message queue, resulting in a heap Use-After-Free (UAF) within the hal_nfc_default process.

To resolve this, the fix explicitly zeros out `nxpncihal_ctrl.gDrvCfg.nClientId` immediately before calling `phOsalNfc_Timer_Cleanup()` across all teardown paths in the `pn8x`, `snxxx/halimpl`, and `snxxx/halimpl_v2` HAL variants. Nullifying this client ID ensures that any delayed timer callbacks will encounter a zeroed message queue ID and return early, safely avoiding access to the freed memory structure.

Test: Compiled via `m android.hardware.nfc-service.nxp` and `m`, ran `atest`, and verified with `alint`.

Bug: 522375782
Flag: EXEMPT BUGFIX
This was autogenerated by MendIt (go/androidmendit)

(cherry picked from commit e886a13085a2a1f6fb8c26f758e57510ee322650)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:9dec300be3955f19f592a32f669df0b2b1a9c20f
Merged-In: I375d72e090aef39d9e70f97f50c3073f95227785
Change-Id: I375d72e090aef39d9e70f97f50c3073f95227785
---
 pn8x/halimpl/hal/phNxpNciHal.cc     | 14 ++++++++++++--
 snxxx/halimpl/hal/phNxpNciHal.cc    | 11 +++++++++--
 snxxx/halimpl_v2/hal/phNxpNciHal.cc |  3 +++
 3 files changed, 24 insertions(+), 4 deletions(-)

diff --git a/pn8x/halimpl/hal/phNxpNciHal.cc b/pn8x/halimpl/hal/phNxpNciHal.cc
index a246e52..e3c39d9 100644
--- a/pn8x/halimpl/hal/phNxpNciHal.cc
+++ b/pn8x/halimpl/hal/phNxpNciHal.cc
@@ -385,8 +385,11 @@ static NFCSTATUS phNxpNciHal_fw_download(void) {
   if (status != NFCSTATUS_SUCCESS) {
     if (NFCSTATUS_SUCCESS != phNxpNciHal_fw_mw_ver_check()) {
       NXPLOG_NCIHAL_D("Chip Version Middleware Version mismatch!!!!");
+      intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+      nxpncihal_ctrl.gDrvCfg.nClientId = 0;
       phOsalNfc_Timer_Cleanup();
       phTmlNfc_Shutdown_CleanUp();
+      phDal4Nfc_msgrelease(client_id);
       status = NFCSTATUS_FAILED;
     } else {
       NXPLOG_NCIHAL_E("FW download failed, Continue NFCC init");
@@ -2066,8 +2069,11 @@ NFCSTATUS phNxpNciHalRFConfigCmdRecSequence() {
           (pphTmlNfc_TransactCompletionCb_t)&phNxpNciHal_read_complete, NULL);
       if (status != NFCSTATUS_PENDING) {
         NXPLOG_NCIHAL_E("TML Read status error status = %x", status);
+        intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+        nxpncihal_ctrl.gDrvCfg.nClientId = 0;
         phOsalNfc_Timer_Cleanup();
         phTmlNfc_Shutdown();
+        phDal4Nfc_msgrelease(client_id);
         status = NFCSTATUS_FAILED;
       }
       break;
@@ -2185,6 +2191,8 @@ int phNxpNciHal_close(bool bShutdown) {
     status = phTmlNfc_ReadAbort();
     status = phTmlNfc_WriteAbort();
 
+    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2195,7 +2203,7 @@ int phNxpNciHal_close(bool bShutdown) {
 
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
+    phDal4Nfc_msgrelease(client_id);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
 
@@ -2235,6 +2243,8 @@ int phNxpNciHal_Minclose(void) {
     status = phTmlNfc_ReadAbort();
     status = phTmlNfc_WriteAbort();
 
+    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2245,7 +2255,7 @@ int phNxpNciHal_Minclose(void) {
 
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
+    phDal4Nfc_msgrelease(client_id);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
 
diff --git a/snxxx/halimpl/hal/phNxpNciHal.cc b/snxxx/halimpl/hal/phNxpNciHal.cc
index 2e0edd8..6e816cf 100644
--- a/snxxx/halimpl/hal/phNxpNciHal.cc
+++ b/snxxx/halimpl/hal/phNxpNciHal.cc
@@ -490,9 +490,12 @@ static NFCSTATUS phNxpNciHal_force_fw_download(uint8_t seq_handler_offset,
       fw_download_success = TRUE;
     } else if (status == NFCSTATUS_FW_CHECK_INTEGRITY_FAILED ||
                (phNxpNciHal_fw_mw_ver_check() != NFCSTATUS_SUCCESS)) {
+      intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+      nxpncihal_ctrl.gDrvCfg.nClientId = 0;
       phOsalNfc_Timer_Cleanup();
       phNxpTempMgr::GetInstance().Reset();
       phTmlNfc_Shutdown_CleanUp();
+      phDal4Nfc_msgrelease(client_id);
       return NFCSTATUS_CMD_ABORTED;
     }
 
@@ -2567,6 +2570,8 @@ close_and_return:
     status = phTmlNfc_ReadAbort();
     status = phTmlNfc_WriteAbort();
 
+    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2578,7 +2583,7 @@ close_and_return:
     phNxpTempMgr::GetInstance().Reset();
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
+    phDal4Nfc_msgrelease(client_id);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
 
@@ -2648,6 +2653,8 @@ void phNxpNciHal_clean_resources() {
     if (status != NFCSTATUS_SUCCESS) {
       NXPLOG_TML_E("phTmlNfc_ReadAbort Failed");
     }
+    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
+    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2663,7 +2670,7 @@ void phNxpNciHal_clean_resources() {
     phNxpTempMgr::GetInstance().Reset();
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
+    phDal4Nfc_msgrelease(client_id);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
   }
diff --git a/snxxx/halimpl_v2/hal/phNxpNciHal.cc b/snxxx/halimpl_v2/hal/phNxpNciHal.cc
index dd965b7..afe6e88 100644
--- a/snxxx/halimpl_v2/hal/phNxpNciHal.cc
+++ b/snxxx/halimpl_v2/hal/phNxpNciHal.cc
@@ -321,6 +321,7 @@ static NFCSTATUS phNxpNciHal_force_fw_download(uint8_t seq_handler_offset,
       fw_download_success = true;
     } else if (status == NFCSTATUS_FW_CHECK_INTEGRITY_FAILED ||
                (phNxpNciHal_fw_mw_ver_check() != NFCSTATUS_SUCCESS)) {
+      nxpncihal_ctrl.gDrvCfg.nClientId = 0;
       phOsalNfc_Timer_Cleanup();
       phNxpTempMgr::GetInstance().Reset();
       phTmlNfc_Shutdown_CleanUp();
@@ -2304,6 +2305,7 @@ close_and_return:
     phNxpNciHal_complete(NFCSTATUS_SUCCESS, PHNXP_NCIHAL_OP_CLOSE);
     /* Abort any pending read and write */
     (void)phTmlNfc_ReadAbort();
+    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     (void)phTmlNfc_Shutdown();
@@ -2365,6 +2367,7 @@ void phNxpNciHal_clean_resources() {
     if (status != NFCSTATUS_SUCCESS) {
       NXPLOG_TML_E("phTmlNfc_ReadAbort Failed");
     }
+    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
-- 
2.55.0

